Dnn.AdminExperience/Dnn.PersonaBar.Extensions/Services/SecurityController.cs

```diff
@@ -1023,6 +1027,283 @@ public HttpResponseMessage GetLastModifiedSettings()
+                response = ApiTokenController.Instance.ApiTokenKeyList(ApiTokenScope.Portal, Thread.CurrentThread.CurrentUICulture.Name);
+            }
+            else
+            {
+                // If the user is regular, set the API token scope to user.
+                response = ApiTokenController.Instance.ApiTokenKeyList(ApiTokenScope.User, Thread.CurrentThread.CurrentUICulture.Name);
+            }
+
+            // Returns the response.
+            return this.Request.CreateResponse(HttpStatusCode.OK, response.Values);
+        }
+
+        /// <summary>
+        /// Creates a new <see cref="ApiToken"/> object with the specified parameters and returns it.
+        /// </summary>
+        /// <param name="data">The parameters for the creation of this token.</param>
+        /// <returns>A new <see cref="ApiToken"/> object.</returns>
+        [HttpPost]
+        [AdvancedPermission(MenuName = Components.Constants.MenuName, Permission = Components.Constants.ManageApiTokens)]
+
+        public HttpResponseMessage CreateApiToken(CreateApiTokenRequest data)
+        {
+            var settings = ApiTokenSettings.GetSettings(this.PortalId);
+            if (!settings.ApiTokensEnabled || (!settings.AllowApiTokens && data.Scope != (int)ApiTokenScope.Host))
+            {
+                return this.Request.CreateErrorResponse(HttpStatusCode.BadRequest, "API tokens are disabled");
+            }
+
+            var requestedScope = ApiTokenScope.User;
+
+            // Checks if the user is authorized.
+            var user = this.UserInfo;
+            if (user.IsSuperUser)
+            {
+                requestedScope = (ApiTokenScope)data.Scope;
+            }
+            else if (user.IsAdmin)
+            {
+                if (data.Scope > 1)
+                {
+                    return this.Request.CreateErrorResponse(HttpStatusCode.BadRequest, "Invalid scope");
+                }
+                else
+                {
+                    requestedScope = (ApiTokenScope)data.Scope;
+                }
+            }
+            else
+            {
+                if (data.Scope > 1)
+                {
+                    return this.Request.CreateErrorResponse(HttpStatusCode.BadRequest, "Invalid scope");
+                }
+            }
+
+            // Check the expiration time
+            var requestedExpiration = DateTime.Parse(data.ExpiresOn);
+            var maxExpirationTime = DateTime.Now;
+            switch (settings.MaximumTimespanMeasure)
+            {
+                case "y":
+                    maxExpirationTime = maxExpirationTime.AddYears(settings.MaximumTimespan);
+                    break;
+                case "m":
+                    maxExpirationTime = maxExpirationTime.AddMonths(settings.MaximumTimespan);
+                    break;
+                case "d":
+                    maxExpirationTime = maxExpirationTime.AddDays(settings.MaximumTimespan);
+                    break;
+            }

```

Comment:
Does this need to handle `"w"` (and should `"m"` be uppercase)?

Reply:
Yes. And regarding lower/upper case: these values correspond to how Dayjs names things. It made front end code a little bit easier. But we can do all lower case if we want.

Reply:
BTW now outdated